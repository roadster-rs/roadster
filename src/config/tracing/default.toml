[tracing]
format = "compact"
trace-propagation = true
