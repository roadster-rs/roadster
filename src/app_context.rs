use crate::app::App;
use crate::config::app_config::AppConfig;
#[cfg(feature = "db-sql")]
use sea_orm::Database;
#[cfg(feature = "db-sql")]
use sea_orm::DatabaseConnection;
use std::sync::Arc;
#[cfg(feature = "sidekiq")]
use tracing::info;

#[derive(Clone)]
pub struct AppContext<T = ()> {
    inner: Arc<AppContextInner>,
    custom: Arc<T>,
}

impl<T> AppContext<T> {
    // This method isn't used when running tests; only the mocked version is used.
    #[cfg_attr(test, allow(dead_code))]
    // The `A` type parameter isn't used in some feature configurations
    #[allow(clippy::extra_unused_type_parameters)]
    pub(crate) async fn new<A>(config: AppConfig) -> anyhow::Result<AppContext<()>>
    where
        A: App,
    {
        #[cfg(feature = "db-sql")]
        let db = Database::connect(A::db_connection_options(&config)?).await?;

        #[cfg(feature = "sidekiq")]
        let (redis_enqueue, redis_fetch) = {
            let sidekiq_config = &config.service.sidekiq;
            let redis_config = &sidekiq_config.custom.redis;
            let redis = sidekiq::RedisConnectionManager::new(redis_config.uri.to_string())?;
            let redis_enqueue = {
                let pool = bb8::Pool::builder().min_idle(redis_config.enqueue_pool.min_idle);
                let pool = redis_config
                    .enqueue_pool
                    .max_connections
                    .iter()
                    .fold(pool, |pool, max_conns| pool.max_size(*max_conns));
                pool.build(redis.clone()).await?
            };
            let redis_fetch = if redis_config
                .fetch_pool
                .max_connections
                .iter()
                .any(|max_conns| *max_conns == 0)
            {
                info!(
                "Redis fetch pool configured with size of zero, will not start the Sidekiq processor"
            );
                None
            } else {
                let pool = bb8::Pool::builder().min_idle(redis_config.fetch_pool.min_idle);
                let pool = redis_config
                    .fetch_pool
                    .max_connections
                    .iter()
                    .fold(pool, |pool, max_conns| pool.max_size(*max_conns));
                Some(pool.build(redis.clone()).await?)
            };
            (redis_enqueue, redis_fetch)
        };

        let inner = AppContextInner {
            config,
            #[cfg(feature = "db-sql")]
            db,
            #[cfg(feature = "sidekiq")]
            redis_enqueue,
            #[cfg(feature = "sidekiq")]
            redis_fetch,
        };
        Ok(AppContext {
            inner: Arc::new(inner),
            custom: Arc::new(()),
        })
    }

    pub fn with_custom<NewT: 'static>(self, custom: NewT) -> AppContext<NewT> {
        AppContext {
            inner: self.inner,
            custom: Arc::new(custom),
        }
    }

    pub fn config(&self) -> &AppConfig {
        &self.inner.config
    }

    #[cfg(feature = "db-sql")]
    pub fn db(&self) -> &DatabaseConnection {
        &self.inner.db
    }

    #[cfg(feature = "sidekiq")]
    pub fn redis_enqueue(&self) -> &sidekiq::RedisPool {
        &self.inner.redis_enqueue
    }

    #[cfg(feature = "sidekiq")]
    pub fn redis_fetch(&self) -> &Option<sidekiq::RedisPool> {
        &self.inner.redis_fetch
    }

    pub fn custom(&self) -> &T {
        &self.custom
    }
}

#[derive(Debug)]
struct AppContextInner {
    config: AppConfig,
    #[cfg(feature = "db-sql")]
    db: DatabaseConnection,
    #[cfg(feature = "sidekiq")]
    redis_enqueue: sidekiq::RedisPool,
    /// The Redis connection pool used by [sidekiq::Processor] to fetch Sidekiq jobs from Redis.
    /// May be `None` if the [fetch_pool.max_connections][crate::config::service::worker::sidekiq::ConnectionPool]
    /// config is set to zero, in which case the [sidekiq::Processor] would also not be started.
    #[cfg(feature = "sidekiq")]
    redis_fetch: Option<sidekiq::RedisPool>,
}

#[cfg(test)]
mockall::mock! {
    pub AppContext<T: 'static = ()> {
        pub fn config(&self) -> &AppConfig;

        #[cfg(feature = "db-sql")]
        pub fn db(&self) -> &DatabaseConnection;

        #[cfg(feature = "sidekiq")]
        pub fn redis_enqueue(&self) -> &sidekiq::RedisPool;

        #[cfg(feature = "sidekiq")]
        pub fn redis_fetch(&self) -> &Option<sidekiq::RedisPool>;

        pub fn with_custom<NewT: 'static>(self, custom: NewT) -> MockAppContext<NewT>;
    }

    impl<T> Clone for AppContext<T> {
        fn clone(&self) -> Self;
    }
}
