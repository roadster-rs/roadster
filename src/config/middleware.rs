use crate::app_context::AppContext;
use crate::controller::middleware::request_id::{PropagateRequestIdConfig, SetRequestIdConfig};
use crate::controller::middleware::sensitive_headers::{
    SensitiveRequestHeadersConfig, SensitiveResponseHeadersConfig,
};
use crate::controller::middleware::tracing::TracingConfig;
use serde_derive::{Deserialize, Serialize};

pub const PRIORITY_FIRST: i32 = -10_000;
pub const PRIORITY_LAST: i32 = 10_000;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct Middleware {
    pub default_enable: bool,
    pub sensitive_request_headers: MiddlewareConfig<SensitiveRequestHeadersConfig>,
    pub sensitive_response_headers: MiddlewareConfig<SensitiveResponseHeadersConfig>,
    pub set_request_id: MiddlewareConfig<SetRequestIdConfig>,
    pub propagate_request_id: MiddlewareConfig<PropagateRequestIdConfig>,
    pub tracing: MiddlewareConfig<TracingConfig>,
}

impl Default for Middleware {
    fn default() -> Self {
        // Before request middlewares
        let mut priority = PRIORITY_FIRST;
        let sensitive_request_headers: MiddlewareConfig<SensitiveRequestHeadersConfig> =
            Default::default();
        let sensitive_request_headers = sensitive_request_headers.set_priority(priority);

        priority += 10;
        let set_request_id: MiddlewareConfig<SetRequestIdConfig> = Default::default();
        let set_request_id = set_request_id.set_priority(priority);

        // Somewhere in the middle, order doesn't particularly matter
        let tracing: MiddlewareConfig<TracingConfig> = Default::default();

        // Before response middlewares
        let mut priority = PRIORITY_LAST;
        let sensitive_response_headers: MiddlewareConfig<SensitiveResponseHeadersConfig> =
            Default::default();
        let sensitive_response_headers = sensitive_response_headers.set_priority(priority);

        priority -= 10;
        let propagate_request_id: MiddlewareConfig<PropagateRequestIdConfig> = Default::default();
        let propagate_request_id = propagate_request_id.set_priority(priority);

        Self {
            default_enable: true,
            sensitive_request_headers,
            sensitive_response_headers,
            set_request_id,
            propagate_request_id,
            tracing,
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct CommonConfig {
    // Optional so we can tell the difference between a consumer explicitly enabling/disabling
    // the middleware, vs the middleware being enabled/disabled by default.
    // If this is `None`, the value will match the value of `Middleware#default_enable`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable: Option<bool>,
    pub priority: i32,
}

impl CommonConfig {
    pub fn set_priority(mut self, priority: i32) -> Self {
        self.priority = priority;
        self
    }

    pub fn enabled(&self, context: &AppContext) -> bool {
        self.enable
            .unwrap_or(context.config.middleware.default_enable)
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct MiddlewareConfig<T: Default> {
    #[serde(flatten)]
    pub common: CommonConfig,
    #[serde(flatten)]
    pub custom: T,
}

impl<T: Default> MiddlewareConfig<T> {
    pub fn set_priority(mut self, priority: i32) -> Self {
        self.common = self.common.set_priority(priority);
        self
    }
}
