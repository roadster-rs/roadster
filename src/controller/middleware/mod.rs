pub mod default;
pub mod request_id;
pub mod sensitive_headers;
pub mod tracing;

use crate::app_context::AppContext;
use axum::Router;

pub trait Middleware {
    fn name(&self) -> String;
    fn enabled(&self, context: &AppContext) -> bool;
    fn priority(&self, context: &AppContext) -> i32;
    fn install(&self, router: Router, context: &AppContext) -> anyhow::Result<Router>;
}
