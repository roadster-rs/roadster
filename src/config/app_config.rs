#[cfg(feature = "db-sql")]
use std::time::Duration;

use anyhow::anyhow;
use config::{Case, Config};
use dotenvy::dotenv;
use serde_derive::{Deserialize, Serialize};
#[cfg(feature = "db-sql")]
use serde_with::serde_as;
use url::Url;

use crate::config::environment::{Environment, ENVIRONMENT_ENV_VAR_NAME};
use crate::config::initializer::Initializer;
use crate::config::middleware::Middleware;
use crate::util::serde_util::{default_true, UriOrString};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct AppConfig {
    pub app: App,
    pub server: Server,
    pub auth: Auth,
    pub tracing: Tracing,
    pub environment: Environment,
    #[cfg(feature = "db-sql")]
    pub database: Database,
    #[cfg(feature = "sidekiq")]
    pub worker: Worker,
    #[serde(default)]
    pub middleware: Middleware,
    #[serde(default)]
    pub initializer: Initializer,
}

pub const ENV_VAR_PREFIX: &str = "ROADSTER";
pub const ENV_VAR_SEPARATOR: &str = ".";

impl AppConfig {
    // This runs before tracing is initialized, so we need to use `println` in order to
    // log from this method.
    #[allow(clippy::disallowed_macros)]
    pub fn new(environment: Option<Environment>) -> anyhow::Result<Self> {
        dotenv().ok();

        let environment = if let Some(environment) = environment {
            println!("Using environment from CLI args: {environment:?}");
            environment
        } else {
            Environment::new()?
        };
        let environment_str: &str = environment.into();

        let config: AppConfig = Config::builder()
            .add_source(config::File::with_name("config/default.toml"))
            .add_source(config::File::with_name(&format!(
                "config/{environment_str}.toml"
            )))
            .add_source(
                config::Environment::default()
                    .prefix(ENV_VAR_PREFIX)
                    .convert_case(Case::Kebab)
                    .separator(ENV_VAR_SEPARATOR),
            )
            .set_override(ENVIRONMENT_ENV_VAR_NAME, environment_str)?
            .build()?
            .try_deserialize()
            .map_err(|err| anyhow!("Unable to deserialize app config: {err:?}"))?;

        // Validations
        // Todo: Is there a crate that would make this easier?
        debug_assert_eq!(
            config.middleware.set_request_id.custom.common.header_name,
            config
                .middleware
                .propagate_request_id
                .custom
                .common
                .header_name,
            "A different request ID header name is used when handling a request vs when sending a response."
        );

        Ok(config)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct App {
    pub name: String,
    /// Shutdown the whole app if an error occurs in one of the app's top-level tasks (API, workers, etc).
    #[serde(default = "default_true")]
    pub shutdown_on_error: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Server {
    pub host: String,
    pub port: u32,
}

impl Server {
    pub fn url(&self) -> String {
        format!("{}:{}", self.host, self.port)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Auth {
    pub jwt: Jwt,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Jwt {
    pub secret: String,
    #[serde(default)]
    pub claims: JwtClaims,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct JwtClaims {
    // Todo: Default to the server URL?
    pub audience: Vec<UriOrString>,
    /// Claim names to require, in addition to the default-required `exp` claim.
    pub required_claims: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Tracing {
    pub level: String,
    /// The name of the service to use for the OpenTelemetry `service.name` field. If not provided,
    /// will use the [`App::name`][App] config value, translated to `snake_case`.
    pub service_name: Option<String>,
    /// Propagate traces across service boundaries. Mostly useful in microservice architectures.
    #[serde(default = "default_true")]
    pub trace_propagation: bool,
    /// URI of the OTLP exporter where traces/metrics/logs will be sent.
    pub otlp_endpoint: Option<Url>,
}

#[cfg(feature = "db-sql")]
#[serde_as]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Database {
    /// This can be overridden with an environment variable, e.g. `ROADSTER.DATABASE.URI=postgres://example:example@example:1234/example_app`
    pub uri: Url,
    /// Whether to automatically apply migrations during the app's start up. Migrations can also
    /// be manually performed via the `roadster migration [COMMAND]` CLI command.
    pub auto_migrate: bool,
    #[serde(default = "Database::default_connect_timeout")]
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub connect_timeout: Duration,
    #[serde(default = "Database::default_acquire_timeout")]
    #[serde_as(as = "serde_with::DurationMilliSeconds")]
    pub acquire_timeout: Duration,
    #[serde_as(as = "Option<serde_with::DurationSeconds>")]
    pub idle_timeout: Option<Duration>,
    #[serde_as(as = "Option<serde_with::DurationSeconds>")]
    pub max_lifetime: Option<Duration>,
    #[serde(default)]
    pub min_connections: u32,
    pub max_connections: u32,
}

#[cfg(feature = "db-sql")]
impl Database {
    fn default_connect_timeout() -> Duration {
        Duration::from_millis(1000)
    }

    fn default_acquire_timeout() -> Duration {
        Duration::from_millis(1000)
    }
}

#[cfg(feature = "sidekiq")]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Worker {
    // Todo: Make Redis optional for workers?
    #[cfg(feature = "sidekiq")]
    pub sidekiq: Sidekiq,
}

#[cfg(feature = "sidekiq")]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Sidekiq {
    // Todo: Make Redis optional for workers?
    pub redis: Redis,
    #[serde(default)]
    pub queue_names: Vec<String>,
}

#[cfg(feature = "sidekiq")]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Redis {
    pub uri: Url,
    #[serde(default = "Redis::default_min_idle")]
    pub min_idle: Option<u32>,
    #[serde(default = "Redis::default_max_connections")]
    pub max_connections: u32,
}

#[cfg(feature = "sidekiq")]
impl Redis {
    fn default_min_idle() -> Option<u32> {
        Some(5)
    }

    fn default_max_connections() -> u32 {
        (num_cpus::get() + 5) as u32
    }
}
