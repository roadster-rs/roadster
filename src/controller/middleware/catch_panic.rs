use crate::app_context::AppContext;
use crate::controller::middleware::Middleware;
use axum::Router;
use serde_derive::{Deserialize, Serialize};
use tower_http::catch_panic::CatchPanicLayer;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct CatchPanicConfig {}

pub struct CatchPanicMiddleware;
impl Middleware for CatchPanicMiddleware {
    fn name(&self) -> String {
        "catch-panic".to_string()
    }

    fn enabled(&self, context: &AppContext) -> bool {
        context
            .config
            .middleware
            .catch_panic
            .common
            .enabled(context)
    }

    fn priority(&self, context: &AppContext) -> i32 {
        context.config.middleware.catch_panic.common.priority
    }

    fn install(&self, router: Router, _context: &AppContext) -> anyhow::Result<Router> {
        let router = router.layer(CatchPanicLayer::new());

        Ok(router)
    }
}
