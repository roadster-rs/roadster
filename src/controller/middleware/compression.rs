use crate::app_context::AppContext;
use crate::controller::middleware::Middleware;
use axum::Router;
use serde_derive::{Deserialize, Serialize};
use tower_http::compression::CompressionLayer;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct CompressionConfig {}

pub struct CompressionMiddleware;
impl Middleware for CompressionMiddleware {
    fn name(&self) -> String {
        "compression".to_string()
    }

    fn enabled(&self, context: &AppContext) -> bool {
        context
            .config
            .middleware
            .compression
            .common
            .enabled(context)
    }

    fn priority(&self, context: &AppContext) -> i32 {
        context.config.middleware.compression.common.priority
    }

    fn install(&self, router: Router, _context: &AppContext) -> anyhow::Result<Router> {
        let router = router.layer(CompressionLayer::new());

        Ok(router)
    }
}
