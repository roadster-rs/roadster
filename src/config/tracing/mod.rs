#[cfg(feature = "otel")]
use crate::util::serde_util::default_true;
use config::{FileFormat, FileSourceString};
use serde_derive::{Deserialize, Serialize};
#[cfg(feature = "otel")]
use url::Url;
use validator::Validate;

pub fn default_config() -> config::File<FileSourceString, FileFormat> {
    config::File::from_str(include_str!("default.toml"), FileFormat::Toml)
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Tracing {
    pub level: String,

    /// The name of the service to use for the OpenTelemetry `service.name` field. If not provided,
    /// will use the [`App::name`][crate::config::app_config::App] config value, translated to `snake_case`.
    #[cfg(feature = "otel")]
    pub service_name: Option<String>,

    /// Propagate traces across service boundaries. Mostly useful in microservice architectures.
    #[serde(default = "default_true")]
    #[cfg(feature = "otel")]
    pub trace_propagation: bool,

    /// URI of the OTLP exporter where traces/metrics/logs will be sent.
    #[cfg(feature = "otel")]
    pub otlp_endpoint: Option<Url>,
}

// To simplify testing, these are only run when all of the config fields are available
#[cfg(all(test, feature = "otel"))]
mod deserialize_tests {
    use super::*;
    use crate::util::test_util::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        level = "debug"
        "#
    )]
    #[case(
        r#"
        level = "info"
        service-name = "foo"
        "#
    )]
    #[case(
        r#"
        level = "error"
        trace-propagation = false
        "#
    )]
    #[case(
        r#"
        level = "debug"
        otlp-endpoint = "https://example.com:1234"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn sidekiq(_case: TestCase, #[case] config: &str) {
        let tracing: Tracing = toml::from_str(config).unwrap();

        assert_toml_snapshot!(tracing);
    }
}
