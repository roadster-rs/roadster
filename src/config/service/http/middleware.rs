#[mockall_double::double]
use crate::app_context::AppContext;
use crate::config::app_config::CustomConfig;
use crate::service::http::middleware::catch_panic::CatchPanicConfig;
use crate::service::http::middleware::compression::{
    RequestDecompressionConfig, ResponseCompressionConfig,
};
use crate::service::http::middleware::request_id::{PropagateRequestIdConfig, SetRequestIdConfig};
use crate::service::http::middleware::sensitive_headers::{
    SensitiveRequestHeadersConfig, SensitiveResponseHeadersConfig,
};
use crate::service::http::middleware::size_limit::SizeLimitConfig;
use crate::service::http::middleware::timeout::TimeoutConfig;
use crate::service::http::middleware::tracing::TracingConfig;
use serde_derive::{Deserialize, Serialize};
use std::collections::BTreeMap;

pub const PRIORITY_FIRST: i32 = -10_000;
pub const PRIORITY_LAST: i32 = 10_000;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct Middleware {
    pub default_enable: bool,
    pub sensitive_request_headers: MiddlewareConfig<SensitiveRequestHeadersConfig>,
    pub sensitive_response_headers: MiddlewareConfig<SensitiveResponseHeadersConfig>,
    pub set_request_id: MiddlewareConfig<SetRequestIdConfig>,
    pub propagate_request_id: MiddlewareConfig<PropagateRequestIdConfig>,
    pub tracing: MiddlewareConfig<TracingConfig>,
    pub catch_panic: MiddlewareConfig<CatchPanicConfig>,
    pub response_compression: MiddlewareConfig<ResponseCompressionConfig>,
    pub request_decompression: MiddlewareConfig<RequestDecompressionConfig>,
    pub timeout: MiddlewareConfig<TimeoutConfig>,
    pub size_limit: MiddlewareConfig<SizeLimitConfig>,
    /// Allows providing configs for custom middleware. Any configs that aren't pre-defined above
    /// will be collected here.
    ///
    /// # Examples
    ///
    /// ```toml
    /// [middleware.foo]
    /// enable = true
    /// priority = 10
    /// x = "y"
    /// ```
    ///
    /// This will be parsed as:
    /// ```raw
    /// Middleware#custom: {
    ///     "foo": {
    ///         MiddlewareConfig#common: {
    ///             enable: true,
    ///             priority: 10
    ///         },
    ///         MiddlewareConfig<CustomConfig>#custom: {
    ///             config: {
    ///                 "x": "y"
    ///             }
    ///         }
    ///     }
    /// }
    /// ```
    #[serde(flatten)]
    pub custom: BTreeMap<String, MiddlewareConfig<CustomConfig>>,
}

impl Default for Middleware {
    fn default() -> Self {
        // Before request middlewares
        let mut priority = PRIORITY_FIRST;
        let sensitive_request_headers: MiddlewareConfig<SensitiveRequestHeadersConfig> =
            Default::default();
        let sensitive_request_headers = sensitive_request_headers.set_priority(priority);

        priority += 10;
        let set_request_id: MiddlewareConfig<SetRequestIdConfig> = Default::default();
        let set_request_id = set_request_id.set_priority(priority);

        priority += 10;
        let tracing: MiddlewareConfig<TracingConfig> = Default::default();
        let tracing = tracing.set_priority(priority);

        priority += 10;
        let size_limit: MiddlewareConfig<SizeLimitConfig> = Default::default();
        let size_limit = size_limit.set_priority(priority);

        priority += 10;
        let request_decompression: MiddlewareConfig<RequestDecompressionConfig> =
            Default::default();
        let request_decompression = request_decompression.set_priority(priority);

        // Somewhere in the middle, order doesn't particularly matter
        let catch_panic: MiddlewareConfig<CatchPanicConfig> = Default::default();
        let response_compression: MiddlewareConfig<ResponseCompressionConfig> = Default::default();
        let timeout: MiddlewareConfig<TimeoutConfig> = Default::default();

        // Before response middlewares
        let mut priority = PRIORITY_LAST;
        let sensitive_response_headers: MiddlewareConfig<SensitiveResponseHeadersConfig> =
            Default::default();
        let sensitive_response_headers = sensitive_response_headers.set_priority(priority);

        priority -= 10;
        let propagate_request_id: MiddlewareConfig<PropagateRequestIdConfig> = Default::default();
        let propagate_request_id = propagate_request_id.set_priority(priority);

        Self {
            default_enable: true,
            sensitive_request_headers,
            sensitive_response_headers,
            set_request_id,
            propagate_request_id,
            tracing,
            catch_panic,
            response_compression,
            request_decompression,
            timeout,
            size_limit,
            custom: Default::default(),
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct CommonConfig {
    // Optional so we can tell the difference between a consumer explicitly enabling/disabling
    // the middleware, vs the middleware being enabled/disabled by default.
    // If this is `None`, the value will match the value of `Middleware#default_enable`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable: Option<bool>,
    pub priority: i32,
}

impl CommonConfig {
    pub fn set_priority(mut self, priority: i32) -> Self {
        self.priority = priority;
        self
    }

    pub fn enabled<S>(&self, context: &AppContext<S>) -> bool {
        self.enable.unwrap_or(
            context
                .config()
                .service
                .http
                .custom
                .middleware
                .default_enable,
        )
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
pub struct MiddlewareConfig<T: Default> {
    #[serde(flatten)]
    pub common: CommonConfig,
    #[serde(flatten)]
    pub custom: T,
}

impl<T: Default> MiddlewareConfig<T> {
    pub fn set_priority(mut self, priority: i32) -> Self {
        self.common = self.common.set_priority(priority);
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app_context::MockAppContext;
    use crate::config::app_config::AppConfig;
    use rstest::rstest;
    use serde_json::Value;

    #[rstest]
    #[case(true, None, true)]
    #[case(true, Some(true), true)]
    #[case(true, Some(false), false)]
    #[case(false, None, false)]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn common_config_enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::empty(None).unwrap();
        config.service.http.custom.middleware.default_enable = default_enable;

        let mut context = MockAppContext::<()>::default();
        context.expect_config().return_const(config);

        let common_config = CommonConfig {
            enable,
            ..Default::default()
        };

        // Act/Assert
        assert_eq!(common_config.enabled(&context), expected_enabled);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn custom_config() {
        // Note: since we're parsing into a Middleware config struct directly, we don't
        // need to prefix `foo` with `middleware`. If we want to actually provide custom middleware
        // configs, the table key will need to be `[middleware.foo]`.
        let config = r#"
        [foo]
        enable = true
        priority = 10
        x = "y"
        "#;
        let config: Middleware = toml::from_str(config).unwrap();

        assert!(config.custom.contains_key("foo"));

        let config = config.custom.get("foo").unwrap();
        assert_eq!(config.common.enable, Some(true));
        assert_eq!(config.common.priority, 10);

        assert!(config.custom.config.contains_key("x"));
        let x = config.custom.config.get("x").unwrap();
        assert_eq!(x, &Value::String("y".to_string()));
    }
}
