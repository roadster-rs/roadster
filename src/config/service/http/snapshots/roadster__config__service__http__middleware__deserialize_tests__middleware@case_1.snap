---
source: src/config/service/http/middleware.rs
expression: middleware
---
default-enable = true

[sensitive-request-headers]
priority = -10000
header-names = [
    'authorization',
    'proxy-authorization',
    'cookie',
    'set-cookie',
]

[sensitive-response-headers]
priority = 10000
header-names = [
    'authorization',
    'proxy-authorization',
    'cookie',
    'set-cookie',
]

[set-request-id]
priority = -9990
header-name = 'request-id'

[propagate-request-id]
priority = 9990
header-name = 'request-id'

[tracing]
priority = -9980

[catch-panic]
priority = 0

[response-compression]
priority = 0

[request-decompression]
priority = -9960

[timeout]
priority = 0
timeout = 10000

[size-limit]
priority = -9970
limit = '5 MB'
